pkg/queues/amqp_provider.go

```diff
@@ -0,0 +1,189 @@
+package queues
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/flightctl/flightctl/pkg/log"
+	"github.com/flightctl/flightctl/pkg/reqid"
+	"github.com/go-chi/chi/v5/middleware"
+	amqp "github.com/rabbitmq/amqp091-go"
+	"github.com/sirupsen/logrus"
+)
+
+type amqpProvider struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	url     string
+	log     logrus.FieldLogger
+	wg      *sync.WaitGroup
+	queues  []*amqpQueue
+	stopped bool
+}
+
+func NewAmqpProvider(url string, log logrus.FieldLogger) Provider {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		<-ctx.Done()
+	}()
+	return &amqpProvider{
+		ctx:    ctx,
+		cancel: cancel,
+		url:    url,
+		log:    log,
+		wg:     &wg,
+	}
+}
+
+func (r *amqpProvider) newQueue(queueName string) (*amqpQueue, error) {
+	var (
+		err        error
+		connection *amqp.Connection
+		channel    *amqp.Channel
+	)
+	if r.stopped {
+		return nil, errors.New("provider is stopped")
+	}
+	defer func() {
+		if err != nil {
+			if channel != nil {
+				channel.Close()
+			}
+			if connection != nil {
+				connection.Close()
+			}
+		}
+	}()
```

Comment:
would it be cleaner to declare the defer statement after you instantiate the variables?

Reply:
OK